pkg/controllers/operationjob/replace/replace.go

```diff
@@ -64,35 +59,38 @@ func (p *PodReplaceHandler) Setup(controller controller.Controller, reconcileMix
 }
 
-func (p *PodReplaceHandler) OperateTarget(ctx context.Context, candidate *OpsCandidate, operationJob *appsv1alpha1.OperationJob) error {
-	if candidate.Pod == nil {
-		return nil
-	}
-
-	// parse replace information from origin pod
-	_, replaceIndicated := candidate.Pod.Labels[appsv1alpha1.PodReplaceIndicationLabelKey]
-	_, replaceByReplaceUpdate := candidate.Pod.Labels[appsv1alpha1.PodReplaceByReplaceUpdateLabelKey]
-	_, replaceNewPodExists := candidate.Pod.Labels[appsv1alpha1.PodReplacePairNewId]
-
-	// label pod to trigger replace
-	replaceTriggered := replaceIndicated || replaceByReplaceUpdate || replaceNewPodExists
-	if !replaceTriggered {
-		patch := client.RawPatch(types.StrategicMergePatchType, []byte(fmt.Sprintf(`{"metadata":{"labels":{"%s":"%v"}}}`, appsv1alpha1.PodReplaceIndicationLabelKey, true)))
-		// add finalizer on origin pod before trigger replace
-		if err := controllerutils.AddFinalizer(ctx, p.client, candidate.Pod, OperationJobReplacePodFinalizer); err != nil {
-			retErr := fmt.Errorf("fail to add %s finalizer to origin pod %s/%s : %s", OperationJobReplacePodFinalizer, candidate.Pod.Namespace, candidate.Pod.Name, err.Error())
-			ojutils.SetOpsStatusError(candidate, ReasonUpdateObjectFailed, retErr.Error())
-			return retErr
+func (p *PodReplaceHandler) OperateTargets(ctx context.Context, candidates []*OpsCandidate, operationJob *appsv1alpha1.OperationJob) error {
+	_, err := controllerutils.SlowStartBatch(len(candidates), controllerutils.SlowStartInitialBatchSize, false, func(i int, _ error) error {
+		candidate := candidates[i]
+		if candidate.Pod == nil {
+			return nil
 		}
-		if err := p.client.Patch(ctx, candidate.Pod, patch); err != nil {
-			retErr := fmt.Errorf("fail to label origin pod %s/%s with replace indicate label by replaceUpdate: %s", candidate.Pod.Namespace, candidate.Pod.Name, err)
-			ojutils.SetOpsStatusError(candidate, ReasonUpdateObjectFailed, retErr.Error())
-			return retErr
-		}
-		p.recorder.Eventf(operationJob, corev1.EventTypeNormal, "ReplaceOriginPod", fmt.Sprintf("Succeeded to trigger originPod %s/%s to replace", operationJob.Namespace, candidate.Pod.Name))
-	}
 
-	return nil
+		// parse replace information from origin pod
+		_, replaceIndicated := candidate.Pod.Labels[appsv1alpha1.PodReplaceIndicationLabelKey]
+		_, replaceByReplaceUpdate := candidate.Pod.Labels[appsv1alpha1.PodReplaceByReplaceUpdateLabelKey]
+		_, replaceNewPodExists := candidate.Pod.Labels[appsv1alpha1.PodReplacePairNewId]
+
+		// label pod to trigger replace
+		replaceTriggered := replaceIndicated || replaceByReplaceUpdate || replaceNewPodExists
+		if !replaceTriggered {
+			patch := client.RawPatch(types.StrategicMergePatchType, []byte(fmt.Sprintf(`{"metadata":{"labels":{"%s":"%v"}}}`, appsv1alpha1.PodReplaceIndicationLabelKey, true)))
+			// add finalizer on origin pod before trigger replace
+			if err := controllerutils.AddFinalizer(ctx, p.client, candidate.Pod, OperationJobReplacePodFinalizer); err != nil {
```

Comment:
Why don't we merge the `AddFinalizer` and `client.Patch from line84` together in one Pod Update operation?

Reply:
Yes, they should be merged.
As AddFinalizer always do update pod, this merge action is possible to be done:
https://github.com/KusionStack/kuperator/pull/279/commits/35b1bbbcf99d1f17a96de4e21074b253fbabf57b